Purchases/SwiftSources/LocalReceiptParsing/Builders/ASN1ContainerBuilder.swift

```diff
@@ -0,0 +1,80 @@
+//
+// Created by Andrés Boedo on 7/29/20.
+// Copyright (c) 2020 Purchases. All rights reserved.
+//
+
+import Foundation
+
+struct ASN1ContainerBuilder {
+
+    func build(fromPayload payload: ArraySlice<UInt8>) throws -> ASN1Container {
+        guard payload.count >= 2,
+              let firstByte = payload.first else { throw ReceiptReadingError.asn1ParsingError }
+        let containerClass = try extractClass(byte: firstByte)
+        let encodingType = try extractEncodingType(byte: firstByte)
+        let containerType = try extractType(byte: firstByte)
+        let length = try extractLength(data: payload.dropFirst())
+        let identifierTotalBytes = 1
+        let metadataBytes = identifierTotalBytes + length.totalBytes
+        
+        guard payload.count - metadataBytes >= Int(length.value) else { throw ReceiptReadingError.asn1ParsingError }
+        
+        let internalPayload = payload.dropFirst(metadataBytes).prefix(Int(length.value))
+        var internalContainers: [ASN1Container] = []
+        if encodingType == .constructed {
+            var currentPayload = internalPayload
+            while (currentPayload.count > 0) {
+                let internalContainer = try build(fromPayload: currentPayload)
+                internalContainers.append(internalContainer)
+                currentPayload = currentPayload.dropFirst(internalContainer.totalBytes)
+            }
+        }
+        return ASN1Container(containerClass: containerClass,
+                             containerType: containerType,
+                             encodingType: encodingType,
+                             length: length,
+                             internalPayload: internalPayload,
+                             internalContainers: internalContainers)
+    }
+}
+
+private extension ASN1ContainerBuilder {
```

Comment:
I usually split up Swift files into: 
- main declaration: holds the type declaration and is the most public one (kinda like an obj-c header)
- private extensions to hold the logic, everything inside is private. 
- extensions with protocols: if the class conforms to a protocol, I declare conformance and implement the related methods in a single extension so it's easier to visualize. 

Only disadvantage to this is that variable declarations must live in the main declaration, so this can sort of break the rule when you have to declare private properties in the main declaration.

Reply:
Thanks for the explanation. I actually asked in the other PR. What's the advantage of making them extensions?